### ISSUE TITLE
Unable to build python on Mac with Mono

### ISSUE THREAD

**Issue by element6** [QuantConnect/Lean#279]
Hi,

Thanks for a great open source initiative. While trying to build the python project, I got this error:

/Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/4.5/Microsoft.Common.targets: Error: Command 'build' exited with code: 127. (QuantConnect.Algorithm.Python)

I am not sure if this is related, but my reference to Microsoft.Scripting has a cross next to it and the error is "Assembly not found for framework Mono/.NET 4.5"
![screenshot 2016-02-09 22 29 13](https://cloud.githubusercontent.com/assets/1396264/12915138/1e1aba4c-cf7d-11e5-81a8-94ec56db1b44.png)

Cheers

**Comment by jaredbroad** [comment]
Hey @element6, make sure you read the readme.md file for the python project: https://github.com/QuantConnect/Lean/blob/master/Algorithm.Python/readme.md

Python requires calling build.sh and installing IronPython separately to LEAN.

If you don't need the python project you can disable it by removing it from the solution.

**Comment by element6** [comment]
Thanks Jared. The build.sh does create a dll file. I enabled python because I want to use python. Is it ready for production use?

anyway, I add the python dll to Launcher's References and got this error. "could not locate 'main' module. Please make sure you have a main.py file in your project.: try re-building algorithm"

Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Lean.Launcher.exe
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.ComponentModel.Composition/4.0.0.0__b77a5c561934e089/System.ComponentModel.Composition.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Lean.Engine.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Common.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Configuration.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Newtonsoft.Json.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Core/4.0.0.0__b77a5c561934e089/System.Core.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System/4.0.0.0__b77a5c561934e089/System.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Logging.dll
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Numerics/4.0.0.0__b77a5c561934e089/System.Numerics.dll [External]
20160210 08:11:05 Trace:: Config.Get(): Configuration key not found. Key: plugin-directory - Using default value: 
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/CloneExtensions.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Fasterflect.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/ICSharpCode.SharpZipLib.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.AWT.WinForms.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Beans.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Charsets.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Corba.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Core.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Jdbc.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Management.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Media.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Misc.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Naming.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Remoting.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Security.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.SwingAWT.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Text.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.Util.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.OpenJDK.XML.API.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.Runtime.JNI.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IKVM.Runtime.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/ImpromptuInterface.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Ionic.Zip.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IronPython.Modules.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IronPython.SQLite.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IronPython.Wpf.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/IronPython.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Krs.Ats.IBNet.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/MathNet.Numerics.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Microsoft.Dynamic.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Microsoft.Scripting.AspNet.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Microsoft.Scripting.Metadata.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/Microsoft.Scripting.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/NodaTime.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Algorithm.CSharp.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Algorithm.Python.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Algorithm.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.AlgorithmFactory.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Api.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Brokerages.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Compression.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Fxcm.dll [External]
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Indicators.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Messaging.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/QuantConnect.Queues.dll
Loaded assembly: /Users/cheng/git/Lean/Launcher/bin/Debug/RestSharp.dll [External]
Loaded assembly: Anonymously Hosted DynamicMethods Assembly [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Xml/4.0.0.0__b77a5c561934e089/System.Xml.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Xml.Linq/4.0.0.0__b77a5c561934e089/System.Xml.Linq.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Windows.Forms/4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Drawing/4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/Accessibility/4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Data/4.0.0.0__b77a5c561934e089/System.Data.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/Microsoft.CSharp/4.0.0.0__b03f5f7f11d50a3a/Microsoft.CSharp.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Xaml/4.0.0.0__b77a5c561934e089/System.Xaml.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Configuration/4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Runtime.Serialization/4.0.0.0__b77a5c561934e089/System.Runtime.Serialization.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Runtime.Remoting/4.0.0.0__b77a5c561934e089/System.Runtime.Remoting.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Web/4.0.0.0__b03f5f7f11d50a3a/System.Web.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.IO.Compression/4.0.0.0__b77a5c561934e089/System.IO.Compression.dll [External]
20160210 08:11:07 Trace:: Config.GetValue(): debug-mode - Using default value: False
20160210 08:11:07 Trace:: Config.Get(): Configuration key not found. Key: data-directory - Using default value: ../../../Data/
20160210 08:11:07 Trace:: Config.Get(): Configuration key not found. Key: version-id - Using default value: 
20160210 08:11:07 Trace:: Engine.Main(): LEAN ALGORITHMIC TRADING ENGINE v2.2.0.2 Mode: DEBUG
20160210 08:11:07 Trace:: Engine.Main(): Started 8:11 AM
20160210 08:11:07 Trace:: Engine.Main(): Memory 81Mb-App  13Mb-Used  16384Mb-Total
20160210 08:11:07 Trace:: Config.Get(): Configuration key not found. Key: job-channel - Using default value: 
20160210 08:11:07 Trace:: Config.GetValue(): job-user-id - Using default value: 2147483647
20160210 08:11:07 Trace:: Config.GetValue(): job-project-id - Using default value: 2147483647
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/Mono.Security/4.0.0.0__0738eb9f132ed756/Mono.Security.dll [External]
20160210 08:11:07 Trace:: JobQueue.NextJob(): Selected QuantConnect.Algorithm.Python.dll
20160210 08:11:07 Trace:: Config.Get(): Configuration key not found. Key: history-provider - Using default value: SubscriptionDataReaderHistoryProvider
20160210 08:11:07 Trace:: JOB HANDLERS: 
20160210 08:11:07 Trace::          DataFeed:     QuantConnect.Lean.Engine.DataFeeds.FileSystemDataFeed
20160210 08:11:07 Trace::          Setup:        QuantConnect.Lean.Engine.Setup.ConsoleSetupHandler
20160210 08:11:07 Trace::          RealTime:     QuantConnect.Lean.Engine.RealTime.BacktestingRealTimeHandler
20160210 08:11:07 Trace::          Results:      QuantConnect.Lean.Engine.Results.BacktestingResultHandler
20160210 08:11:07 Trace::          Transactions: QuantConnect.Lean.Engine.TransactionHandlers.BacktestingTransactionHandler
20160210 08:11:07 Trace::          History:      QuantConnect.Lean.Engine.HistoricalData.SubscriptionDataReaderHistoryProvider
20160210 08:11:07 Trace::          Commands:     QuantConnect.Queues.EmptyCommandQueueHandler
20160210 08:11:07 Trace:: Config.GetValue(): ignore-version-checks - Using default value: False
20160210 08:11:07 Trace:: Config.GetValue(): algorithm-manager-time-loop-maximum - Using default value: 10
Thread started:  #2
Thread started: Result Thread #3
20160210 08:11:07 Trace:: Config.Get(): Configuration key not found. Key: ironpython-location - Using default value: ../ironpython/Lib
Thread started: <Thread Pool> #4
Thread started: <Thread Pool> #5
Thread started: <Thread Pool> #6
20160210 08:11:08 Trace:: Loader.TryCreatePythonAlgorithm(): Loading python assembly: QuantConnect.Algorithm.Python.dll
20160210 08:11:08 Trace:: Loader.TryCreatePythonAlgorithm(): Importing python module...
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/Mono.CSharp/4.0.0.0__0738eb9f132ed756/Mono.CSharp.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.ServiceModel.Internals/0.0.0.0__b77a5c561934e089/System.ServiceModel.Internals.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.Web.ApplicationServices/4.0.0.0__31bf3856ad364e35/System.Web.ApplicationServices.dll [External]
Loaded assembly: /Library/Frameworks/Mono.framework/Versions/4.2.2/lib/mono/gac/System.ComponentModel.DataAnnotations/4.0.0.0__31bf3856ad364e35/System.ComponentModel.DataAnnotations.dll [External]
20160210 08:11:08 ERROR:: Loader.TryCreatePythonAlgorithm():  System.IO.IOException: Could not add reference to assembly QuantConnect.Interfaces
  at IronPython.Runtime.ClrModule.AddReference (IronPython.Runtime.CodeContext context, System.String name) [0x00041] in <filename unknown>:0 
  at IronPython.Runtime.ClrModule.AddReference (IronPython.Runtime.CodeContext context, System.Object reference) [0x00022] in <filename unknown>:0 
  at IronPython.Runtime.ClrModule.AddReference (IronPython.Runtime.CodeContext context, System.Object[] references) [0x00033] in <filename unknown>:0 
  at Microsoft.Scripting.Interpreter.ActionCallInstruction`2[T0,T1].Run (Microsoft.Scripting.Interpreter.InterpretedFrame frame) [0x00000] in <filename unknown>:0 
  at Microsoft.Scripting.Interpreter.Interpreter.Run (Microsoft.Scripting.Interpreter.InterpretedFrame frame) [0x00015] in <filename unknown>:0 
20160210 08:11:08 ERROR:: Engine.Run():  System.Exception: Could not add reference to assembly QuantConnect.Interfaces - could not locate 'main' module. Please make sure you have a main.py file in your project.: try re-building algorithm.
  at QuantConnect.Lean.Engine.Setup.ConsoleSetupHandler.CreateAlgorithmInstance (System.String assemblyPath, Language language) [0x0005c] in /Users/cheng/git/Lean/Engine/Setup/ConsoleSetupHandler.cs:94 
  at QuantConnect.Lean.Engine.Engine.Run (QuantConnect.Packets.AlgorithmNodePacket job, System.String assemblyPath) [0x0013c] in /Users/cheng/git/Lean/Engine/Engine.cs:110 
20160210 08:11:08 Trace:: StateCheck.Ping.Run(): Exited thread.
Thread finished:  #2
20160210 08:11:08 Trace:: Config.GetValue(): send-via-api - Using default value: False
20160210 08:11:08 ERROR:: Algorithm.Initialize() Error: Could not add reference to assembly QuantConnect.Interfaces - could not locate 'main' module. Please make sure you have a main.py file in your project.: try re-building algorithm. Stack Trace:   at QuantConnect.Lean.Engine.Setup.ConsoleSetupHandler.CreateAlgorithmInstance (System.String assemblyPath, Language language) [0x0005c] in /Users/cheng/git/Lean/Engine/Setup/ConsoleSetupHandler.cs:94 
  at QuantConnect.Lean.Engine.Engine.Run (QuantConnect.Packets.AlgorithmNodePacket job, System.String assemblyPath) [0x0013c] in /Users/cheng/git/Lean/Engine/Engine.cs:110 
   at QuantConnect.Lean.Engine.Setup.ConsoleSetupHandler.CreateAlgorithmInstance (System.String assemblyPath, Language language) [0x0005c] in /Users/cheng/git/Lean/Engine/Setup/ConsoleSetupHandler.cs:94 
  at QuantConnect.Lean.Engine.Engine.Run (QuantConnect.Packets.AlgorithmNodePacket job, System.String assemblyPath) [0x0013c] in /Users/cheng/git/Lean/Engine/Engine.cs:110 
20160210 08:11:08 Trace:: BacktestingResultHandler.Run(): Ending Thread...
Thread finished: Result Thread #3
20160210 08:11:08 Trace:: Waiting for threads to exit...
20160210 08:11:08 Trace:: Engine.Run(): Disposing of setup handler...
20160210 08:11:08 Trace:: Engine.Main(): Analysis Completed and Results Posted.
20160210 08:11:08 Trace:: FileSystemDataFeed.Exit(): Exit triggered.
20160210 08:11:08 Trace:: Engine.Main(): Analysis Complete. Press any key to continue.
20160210 08:11:08 Trace:: Engine.Main(): Packet removed from queue: BasicTemplateAlgorithm

**Comment by AlexCatarino** [comment]
Hi @element6.

I made a PR that fix this problem in Windows. Since I don't have a Mac, I tried to check if this fix works on Linux (it's all Mono, so it should be the same), but I couldn't run Lean on Ubuntu 15.10...
You don't need to wait this PR to be merged to make it work, because it's quite simple. Eliminate the following line:
`clr.AddReference("QuantConnect.Interfaces")`

I think that as we don't have QuantConnect.Interfaces in the Algorithm.Python and/or Launcher projects, the program throws an error.

**Comment by jaredbroad** [comment]
Closed in #279

**Comment by olegshulyakov** [comment]
Hi @jaredbroad I have made changes according readme but this have not helped.

I have found the reason. QuantConnect.Algorithm.Python.csproj file have following lines:
```
  <PropertyGroup>
    <PostBuildEvent>build</PostBuildEvent>
  </PropertyGroup>
```
I assume that this is launching built.bat on Windows system.
Thats why following should be done on Mac/Linux:
1. Make build script runnable  `chmod +x build.sh`
2. Change content of project file to:
```
  <PropertyGroup>
    <PostBuildEvent>./build.sh</PostBuildEvent>
  </PropertyGroup>
```

Please add this into readme. I hope it will be helpfull
